src/PolyType/SourceGenModel/SourceGenObjectTypeShape.cs

```diff
@@ -34,23 +35,15 @@ public sealed class SourceGenObjectTypeShape<TObject> : SourceGenTypeShape<TObje
     /// <inheritdoc/>
     public override object? Accept(TypeShapeVisitor visitor, object? state = null) => visitor.VisitObject(this, state);
 
-    IReadOnlyList<IPropertyShape> IObjectTypeShape.Properties => _properties ?? CommonHelpers.ExchangeIfNull(ref _properties, (CreatePropertiesFunc?.Invoke()).AsReadOnlyList());
+    /// <inheritdoc/>
+    public IReadOnlyList<IPropertyShape> Properties => _properties ?? CommonHelpers.ExchangeIfNull(ref _properties, (CreatePropertiesFunc?.Invoke()).AsReadOnlyList());
+
+    [DebuggerBrowsable(DebuggerBrowsableState.Never)]
     private IReadOnlyList<IPropertyShape>? _properties;
 
-    IConstructorShape? IObjectTypeShape.Constructor
-    {
-        get
-        {
-            if (!_isConstructorResolved)
-            {
-                _constructor = CreateConstructorFunc?.Invoke();
-                Volatile.Write(ref _isConstructorResolved, true);
-            }
-
-            return _constructor;
-        }
-    }
-
-    private bool _isConstructorResolved;
+    /// <inheritdoc/>
+    public IConstructorShape? Constructor => _constructor ?? (CreateConstructorFunc is null ? null : CommonHelpers.ExchangeIfNull(ref _constructor, CreateConstructorFunc.Invoke()));
```

Comment:
I don't think this will work because `CreateConstructorFunc` can return null, meaning that it will force a recomputation every time that property is being poked. I think we need to keep the original structure but also write to the `_constructor` field using `ExchangeIfNull`. That was my plan when making that refactoring but it seems I forgot to do it.

Reply:
> because CreateConstructorFunc can return null

Actually, that isn't the case but I don't think we should rely on the source generator doing that.